crypto/dsa/dsa_ameth.c

```diff
@@ -514,6 +520,48 @@ static int dsa_pkey_ctrl(EVP_PKEY *pkey, int op, long arg1, void *arg2)
 
 }
 
+static size_t dsa_pkey_dirty_cnt(const EVP_PKEY *pkey)
+{
+    return pkey->pkey.dsa->dirty_cnt;
+}
+
+static void *dsa_pkey_export_to(const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
+{
+    DSA *dsa = pk->pkey.dsa;
+    OSSL_PARAM_BLD tmpl;
+    const BIGNUM *p = DSA_get0_p(dsa);
+    const BIGNUM *g = DSA_get0_g(dsa);
+    const BIGNUM *q = DSA_get0_q(dsa);
+    const BIGNUM *pub_key = DSA_get0_pub_key(dsa);
+    const BIGNUM *priv_key = DSA_get0_priv_key(dsa);
+    OSSL_PARAM *params;
+    void *provkey = NULL;
+
+    if (p == NULL || g == NULL || pub_key == NULL)
+        return NULL;
+
+    ossl_param_bld_init(&tmpl);
+    if (!ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_DSA_P, p)
+        || !ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_DSA_G, g)
+        || !ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_DSA_Q, q)
```

Comment:
I notice you dont check for null q here? But you do a extra null check in the RSA?

Reply:
Because with RSA, they are at least theoretically not mandatory (only n, e and d are needed for encryption, decryption, signing and verifying), while with DSA, p, g and q are needed to sign and verify...  that's at least how I read the algo descriptions...

Reply:
I think if you look up a few lines you can see that q isnt mandatory either (since you only check for p and g being NULL).

Reply:
I realised, though, that it's quite possible the EVP_PKEY only contains domain parameters, so I should not make pub_key mandatory...

Reply:
good point..

Reply:
It wouldnt fall over though? it would just pass a size of -1..